source/extensions/common/aws/metadata_fetcher.cc

```diff
@@ -0,0 +1,175 @@
+                       .setParentSpan(parent_span)
+                       .setSendXff(false)
+                       .setChildSpanName("AWS Metadata Fetch");
+
+    envoy::config::route::v3::RetryPolicy route_retry_policy;
+    route_retry_policy.mutable_num_retries()->set_value(MAX_RETRIES);
+    route_retry_policy.mutable_per_try_timeout()->CopyFrom(
+        Protobuf::util::TimeUtil::MillisecondsToDuration(TIMEOUT));
+    route_retry_policy.mutable_per_try_idle_timeout()->CopyFrom(
+        Protobuf::util::TimeUtil::MillisecondsToDuration(RETRY_DELAY));
+    route_retry_policy.set_retry_on("5xx,gateway-error,connect-failure,reset,refused-stream");
+
+    options.setRetryPolicy(route_retry_policy);
+    options.setBufferBodyForRetry(true);
+    request_ = thread_local_cluster->httpAsyncClient().send(std::move(messagePtr), *this, options);
+  }
+
+  // HTTP async receive method on success.
+  void onSuccess(const Http::AsyncClient::Request&, Http::ResponseMessagePtr&& response) override {
+    complete_ = true;
+    const uint64_t status_code = Http::Utility::getResponseStatus(response->headers());
+    if (status_code == enumToInt(Http::Code::OK)) {
+      ENVOY_LOG(debug, "{}: fetch AWS Metadata [cluster = {}]: success", __func__, cluster_name_);
+      if (response->body().length() != 0) {
+        const auto body = response->bodyAsString();
+        receiver_->onMetadataSuccess(std::move(body));
+      } else {
+        ENVOY_LOG(debug, "{}: fetch AWS Metadata [cluster = {}]: body is empty", __func__,
+                  cluster_name_);
+        receiver_->onMetadataError(MetadataFetcher::MetadataReceiver::Failure::InvalidMetadata);
+      }
+    } else {
+      if (response->body().length() != 0) {
+        ENVOY_LOG(debug, "{}: fetch AWS Metadata [cluster = {}]: response status code {}, body: {}",
+                  __func__, cluster_name_, status_code, response->bodyAsString());
+      } else {
+        ENVOY_LOG(debug,
+                  "{}: fetch AWS Metadata [cluster = {}]: response status code {}, body is empty",
+                  __func__, cluster_name_, status_code);
+      }
+      receiver_->onMetadataError(MetadataFetcher::MetadataReceiver::Failure::Network);
+    }
+    reset();
+  }
+
+  // HTTP async receive method on failure.
+  void onFailure(const Http::AsyncClient::Request&,
+                 Http::AsyncClient::FailureReason reason) override {
+    ENVOY_LOG(debug, "{}: fetch AWS Metadata [cluster = {}]: network error {}", __func__,
+              cluster_name_, enumToInt(reason));
+    complete_ = true;
+    receiver_->onMetadataError(MetadataFetcher::MetadataReceiver::Failure::Network);
+    reset();
+  }
+
+  // TODO(suniltheta): Add metadata fetch status into the span like it is done on ext_authz filter.
+  void onBeforeFinalizeUpstreamSpan(Tracing::Span&, const Http::ResponseHeaderMap*) override {}
+
+private:
+  Upstream::ClusterManager& cm_;
+  bool complete_{};
```

Comment:
Ah, I think I see, but it's horrible. :)

What we could do to avoid having this redundant value indicating the same thing, is instead of `complete_ = true`, `request_.reset()` - the reason it's not quite as I thought is that a callback is called *then* reset() is called, so if the callback calls cancel() (which doesn't really make sense, but I guess can still happen) then it can hit the `request_ && !complete_` condition while complete_ is true and request_ is not yet reset.

But again, since the only time complete_ is checked is to act like request_ has already been reset in cancel(), and the only time request_ is read from is *also* in cancel(), it seems there is no need for two values representing the same state but getting slightly out of sync.

Reply:
If I try to remove `cancel_` then on destruction of `::MetadataFetcherImpl::~MetadataFetcherImpl` object on every test in `metadata_fetcher_test` the cancel is called one final time which seem to crash.

So this cancel_ is just doing a gate keeping before destruction.

Reply:
is there a way to check if cancel is called already on request_ ? just checking `if (request_)` doesn't seem to clarify that `cancel` is called already.

Reply:
I wasn't suggesting trying to avoid the duplicate cancel() - it should be idempotent to call cancel repeatedly.

Rather, the suggestion was that in the places where currently the code says `complete_ = true`, the code would say `request_.reset()` (to signify the same thing, that the request is done with), and then the behavior should be identical after removing all other instances of `complete_`.

Reply:
The `request_.reset()` is getting called at all the places where `complete_= true` is set by the virtue of call to `reset()`
```
  void reset() {
    request_.reset();
    receiver_.reset();
  }
```

I am struggling with, some unit tests that crash on cleanup where `request_.cancel()` is called on mock `request_` object created at https://github.com/envoyproxy/envoy/blob/e9366f2b34be537b1b30ed17a4fd75b4715b5990/test/extensions/filters/http/common/mock.cc#L11C14-L11C14.

Reply:
That's what I thought at first, but this is the difference:
```
      complete_ = true;
      receiver_->onMetadataError(MetadataFetcher::MetadataReceiver::Failure::MissingConfig);
      reset();
```
If cancel() is called during onMetadataError, then it's called while request_ is still set.
If you change that (and others like it) to
```
      request_.reset();
      receiver_->onMetadataError(MetadataFetcher::MetadataReceiver::Failure::MissingConfig);
      reset();
```
Then `request_.reset()` will make it so MetadataFetcher::cancel() is a no-op the same way `complete_ = true` does now.

(Or: how could `request_.cancel()` be called on a mock `request_` object when there is no longer a reference to that object?)

Reply:
Ack on calling `request_.reset();` in place of `complete_ = true;`

---

Cancel is initiated as part of destruction.
```
~MetadataFetcherImpl() override { cancel(); }
```

I could just make it call reset(), and don't have to worry about `request_.cancel()` crash while running some unit tests.

Reply:
Agree, unless the destructor is happening during one of the callbacks, which even if that's the case seems like it should still be problematic since there's a reset() *after* calling the callback, which would be bad to do on a destroyed thing.

I wouldn't change cancel() to reset() in the destructor since the cancel-on-destroy behavior is likely to be used somewhere.

Reply:
2 new things that I tried

1. Making sure `request_` object (in [code](https://github.com/suniltheta/envoy/blob/remove_libcurl_28_only_required_changes/test/extensions/common/aws/metadata_fetcher_test.cc#L98)) is created before creating `fetcher_` object (in [code](https://github.com/suniltheta/envoy/blob/5433e871202010b7e6e19b1038e71f9848fa0aad/test/extensions/common/aws/metadata_fetcher_test.cc#L84)). Thinking that the order of destruction will be in reverse order of object creation. But didn't seem to help https://gist.github.com/suniltheta/1ceed205fcec6042e8ceefc91db33276 the crash is still thrown on destruction of  `fetcher_` unique_ptr. Ideally the call `request_->cancel()` should not even happen given `if (request_)` should return `false` from reset on earlier callbacks.

2. Tried to do emplace of pointer returned in call to mock send ([code](https://github.com/suniltheta/envoy/blob/5433e871202010b7e6e19b1038e71f9848fa0aad/test/extensions/filters/http/common/mock.cc#L26)). Still face the crash on first unit test.

```
     options.setRetryPolicy(route_retry_policy);
     options.setBufferBodyForRetry(true);
-    request_ = makeOptRefFromPtr(
-        thread_local_cluster->httpAsyncClient().send(std::move(messagePtr), *this, options));
+    request_.emplace(*thread_local_cluster->httpAsyncClient().send(std::move(messagePtr), *this, options));
   }
```

Reply:
Ah, I see, the problem is that the request is completing during send() before `request_` has even been set in the first place, so clearing it is too soon. It would be *slightly* nicer than the current setup to have request_ be a tristate of some sort so it can represent "none", "active request" or "completed request", like an `absl::variant`, but an optional and a bool is probably simpler, so leaving it as-is is reasonable.

Might be nice to have a comment above complete_ to explain why it has to be like this for the next person (I think send() can *potentially* complete or error instantly even when it's not a mock), but I guess complete_ can stay. Sorry for the runaround.

Reply:
I will include a note about the intention to keep `complete_` in my follow up PR, thanks for the approval.